src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/utils.ts

```diff
@@ -441,6 +443,49 @@ export const generateApiLayer = (options: PersistedIndexPatternLayer | TextBased
   };
 };
 
+export const filtersToApiFormat = (filters: Filter[]): LensApiFilterType[] => {
+  return filters.map((filter) => ({
+    language: filter.query?.language,
+    query: filter.query?.query,
+    meta: {},
+  }));
+};
+
+export const queryToApiFormat = (query: Query): LensApiFilterType | undefined => {
+  if (typeof query.query !== 'string') {
+    return;
+  }
+  return {
+    query: query.query,
+    language: query.language as 'kuery' | 'lucene',
+  };
+};
+
+export const filtersToLensState = (filters: LensApiFilterType[]): Filter[] => {
+  return filters.map((filter) => ({
+    query: { query: filter.query, language: filter.language },
+    meta: {},
+  }));
+};
+
+export const queryToLensState = (query: LensApiFilterType): Query => {
+  return query;
+};
+
+export const filtersAndQueryToApiFormat = (state: LensAttributes) => {
+  return {
+    filters: filtersToApiFormat(state.state.filters),
+    query: queryToApiFormat(state.state.query as Query),
+  };
+};
+
+export const filtersAndQueryToLensState = (state: LensApiState) => {
+  return {
+    ...(state.filters ? { filters: filtersToLensState(state.filters) } : {}),
+    ...(state.query ? { query: queryToLensState(state.query as LensApiFilterType) } : {}),
+  };
```

Comment:
It is important to put the ES|QL query at the state root level to make identify the chart correctly as `textBased`.
Without that the ES|QL editor won't show up in the inline edit flyout.

Note that an ES|QL chart cannot have a KQL query so that should be safe at runtime.

```suggestion
  const query =
    state.dataset.type === 'esql'
      ? { query: { esql: state.dataset.query } }
      : state.query
      ? { query: queryToLensState(state.query) }
      : {};
  return {
    ...(state.filters ? { filters: filtersToLensState(state.filters) } : {}),
    ...(query),
```